File: DSA/GreedyAlgo/Code/FractionalKnapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Item {
    int id;
    double weight;
    double value;
    double ratio;
    
    bool operator<(const Item& other) const {
        return ratio > other.ratio;
    }
};

void fractionalKnapsack(vector<Item>& items, double capacity) {
    for (auto& item : items) {
        item.ratio = item.value / item.weight;
    }
    
    sort(items.begin(), items.end());
    
    cout << "Fractional Knapsack Solution:\n";
    double totalValue = 0;
    double usedCapacity = 0;
    
    for (auto& item : items) {
        if (usedCapacity >= capacity) break;
        
        double take = min(capacity - usedCapacity, item.weight);
        cout << "Item " << item.id << ": Take " << take << " units (Weight: " << item.weight 
             << ", Value/Weight: " << item.ratio << ")\n";
        
        totalValue += take * item.ratio;
        usedCapacity += take;
    }
    
    cout << "\nTotal Capacity Used: " << usedCapacity << " / " << capacity << "\n";
    cout << "Total Value: " << totalValue << "\n";
}

int main() {
    cout << "=== Fractional Knapsack (C++) ===\n\n";
    
    vector<Item> items = {
        {1, 10, 60, 0},
        {2, 20, 100, 0},
        {3, 30, 120, 0}
    };
    double capacity = 50;
    
    cout << "Items (ID, Weight, Value):\n";
    for (const auto& item : items) {
        cout << "Item " << item.id << ": Weight " << item.weight << ", Value " << item.value << "\n";
    }
    cout << "Knapsack Capacity: " << capacity << "\n\n";
    
    fractionalKnapsack(items, capacity);
    
    cout << "\n=== Greedy Strategy ===\n";
    cout << "1. Calculate value/weight ratio for each item\n";
    cout << "2. Sort items by ratio (descending)\n";
    cout << "3. For each item in sorted order:\n";
    cout << "   - If item fits completely: take it\n";
    cout << "   - Otherwise: take fraction that fits\n";
    cout << "4. Repeat until capacity full\n\n";
    
    cout << "=== Complexity ===\n";
    cout << "Time Complexity:  O(n log n) for sorting\n";
    cout << "Space Complexity: O(1)\n";
    cout << "Optimal: Yes (greedy works for fractional)\n";
    
    return 0;
}

File: DSA/GreedyAlgo/Code/FractionalKnapsack.cs
// Fractional Knapsack in C#

using System;
using System.Collections.Generic;
using System.Linq;

class FractionalKnapsack {
    class Item : IComparable<Item> {
        public int Id { get; set; }
        public double Weight { get; set; }
        public double Value { get; set; }
        public double Ratio { get; set; }
        
        public Item(int id, double weight, double value) {
            Id = id;
            Weight = weight;
            Value = value;
            Ratio = value / weight;
        }
        
        public int CompareTo(Item other) {
            return other.Ratio.CompareTo(this.Ratio);
        }
    }
    
    public static void FractionalKnapsackSolver(List<Item> items, double capacity) {
        items.Sort();
        
        Console.WriteLine("Fractional Knapsack Solution:");
        double totalValue = 0;
        double usedCapacity = 0;
        
        foreach (Item item in items) {
            if (usedCapacity >= capacity) break;
            
            double take = Math.Min(capacity - usedCapacity, item.Weight);
            Console.WriteLine($"Item {item.Id}: Take {take:F2} units (Weight: {item.Weight}, Value/Weight: {item.Ratio:F2})");
            
            totalValue += take * item.Ratio;
            usedCapacity += take;
        }
        
        Console.WriteLine($"\nTotal Capacity Used: {usedCapacity:F2} / {capacity}");
        Console.WriteLine($"Total Value: {totalValue:F2}");
    }
    
    static void Main() {
        Console.WriteLine("=== Fractional Knapsack (C#) ===\n");
        
        List<Item> items = new List<Item> {
            new Item(1, 10, 60),
            new Item(2, 20, 100),
            new Item(3, 30, 120)
        };
        double capacity = 50;
        
        Console.WriteLine("Items (ID, Weight, Value):");
        foreach (var item in items) {
            Console.WriteLine($"Item {item.Id}: Weight {item.Weight}, Value {item.Value}");
        }
        Console.WriteLine($"Knapsack Capacity: {capacity}\n");
        
        FractionalKnapsackSolver(items, capacity);
        
        Console.WriteLine("\n=== Greedy Strategy ===");
        Console.WriteLine("1. Calculate value/weight ratio for each item");
        Console.WriteLine("2. Sort items by ratio (descending)");
        Console.WriteLine("3. For each item in sorted order:");
        Console.WriteLine("   - If item fits completely: take it");
        Console.WriteLine("   - Otherwise: take fraction that fits");
        Console.WriteLine("4. Repeat until capacity full\n");
        
        Console.WriteLine("=== Complexity ===");
        Console.WriteLine("Time Complexity:  O(n log n) for sorting");
        Console.WriteLine("Space Complexity: O(1)");
        Console.WriteLine("Optimal: Yes (greedy works for fractional)");
    }
}
